src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Query/LaunchProfiles/LaunchSettingsQueryVersionProvider.cs

```diff
@@ -0,0 +1,88 @@
+﻿// Licensed to the .NET Foundation under one or more agreements. The .NET Foundation licenses this file to you under the MIT license. See the LICENSE.md file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.ComponentModel.Composition;
+
+namespace Microsoft.VisualStudio.ProjectSystem.VS.Query
+{
+    /// <summary>
+    /// Aggregates version information from the multiple <see cref="LaunchSettingsTracker" />s
+    /// and passes it on to the bound <see cref="ILaunchSettingsVersionPublisher"/>, if any.
+    /// </summary>
+    /// <remarks>
+    /// The <see cref="LaunchSettingsTracker"/> loads automatically for any project
+    /// supporting launch profiles. However, we don't want to load all of the assemblies
+    /// associated with the Project Query API until they are actually needed, so <see cref="LaunchSettingsTracker"/>
+    /// can't refer to any of them directly. Instead, it communicates with this type,
+    /// which will pass on the version information only after being bound to an <see cref="ILaunchSettingsVersionPublisher"/>.
+    /// </remarks>
+    [Export]
+    internal sealed class LaunchSettingsQueryVersionProvider
+    {
+        private readonly object _lock = new();
+
+        private readonly Dictionary<string, LaunchSettingsTracker> _allLaunchSettingsTrackers = new(StringComparer.Ordinal);
+        private ImmutableDictionary<string, long> _versions = ImmutableStringDictionary<long>.EmptyOrdinal;
+
+        private ILaunchSettingsVersionPublisher? _versionPublisher;
+
+        /// <summary>
+        /// Sets the related <see cref="ILaunchSettingsVersionPublisher"/> and pushes
+        /// the current version information to it.
+        /// </summary>
+        internal void BindToVersionPublisher(ILaunchSettingsVersionPublisher versionPublisher)
+        {
+            lock (_lock)
+            {
+                _versionPublisher = versionPublisher;
+                _versionPublisher.UpdateVersions(_versions);
+            }
+        }
+
+        /// <summary>
+        /// Called when the <paramref name="tracker"/> becomes active.
+        /// </summary>
+        internal void OnLaunchSettingsTrackerActivated(LaunchSettingsTracker tracker)
+        {
+            lock (_lock)
+            {
+                string key = tracker.VersionKey;
+
+                _allLaunchSettingsTrackers.Add(key, tracker);
+                _versions = _versions.SetItem(key, tracker.CurrentVersion);
```

Comment:
Does `SetItem` return a different instance? Is that why you need to reassign? (I notice this pattern applies to other methods too)

Reply:
If we are mutating an immutable dictionary, could we just use a mutable dictionary? I don't know enough about the code here, but I'm assuming we are just letting the original instance get garbage collected (it becomes unreferenced) at some point for every one of these reassignments. Maybe only create immutable dictionaries out of mutable ones when the interfaces (I'm assuming CPS) require an immutable dictionary.